Integration with ESP/CAPS using caps2tacs Wrappers

sean-engelstad opened the pull request:
Summary of Integration with ESP/CAPS and tacsAIM wrapper classes of caps2tacs module
* caps2tacs module provides integration of ESP/CAPS, a parametric CAD geometry tool, for setting up structural analysis problems with a wide range of geometries. 
* Wrapper classes for the tacsAIM reduce the # of lines of code required to setup structural analysis and optimization problems from ESP/CAPS geometries and with TACS structural analysis
* ESP/CAPS does not support MPI by default, these wrapper classes extend the tacsAIM for use in multiproc cases.
* Automatically create a `pyTACS` fea_solver object from a fully setup `TacsAim` wrapper class in caps2tacs module.
* Cascading methods and class method decorator functions enable concise, easy to read source code.
* The user should install ESP/CAPS from the website [https://acdl.mit.edu/ESP/](url) and using the ESPbeta build of ESP/CAPS
 [https://acdl.mit.edu/ESP/archive/ESPbeta.tgz](url).

Classes included in the `caps2tacs` module:
* The builder class `CapsStruct` sets up the `capsProblem` on the root processor, a work directory `capsStruct`, and initializes both the `tacs_aim` for structural analysis input files and `egads_aim` for structural mesh generation. The ESP/CAPS geometry must first be written in a CSM file which stands for Constructive Solid Model of the parametric geometry.
* Wrapper classes `TacsAim` and `EgadsAim` hold the structural analysis inputs and structural meshing respectively. All the following classes have to be registered to the `TacsAim` class before the `pre_analysis()` method can be called to generate structural mesh and structural analysis input files in the nastran .bdf and .dat file format.
* Material objects `Material`, `Isotropic`, and `Orthotropic` help define material properties for groups of elements.
* `ThicknessVariable` objects are built for sizing / thickness optimization corresponding to regions of the ESP/CAPS geometry labeled with the `capsGroup` attributes. If a material is input into the `ThicknessVariable`, it can also build an automatic `ShellProperty` class and register it to the `TacsAim`. 
* `ShapeVarible` objects corresond to ESP/CAPS design parameters to optimize internal structure / shape.
* Constraints and load classes are also available corresponding to ESP/CAPS attributes `capsConstraint` and `capsLoad` respectively. These attributes tell which portions of the geometry receive the constraint / load in the CSM file.
* Finally, the `TacsAim` class must be setup and the pre_analysis ran to generate structural analysis input files, with a property fea_solver available to build a pyTACS object and setup a structural analysis in TACS.

4 examples are written to demonstrate the capabilities of caps2tacs module to build structural geometries with ESP/CAPS from a CSM file and setup TACS structural analysis problems:

1. Steady elastic structural analysis of a NACA wing with TACS `StaticProblem` class.
2. Unsteady elastic structural analysis of NACA wing with TACS `TransientProblem` class.
3. Sizing / panel thickness optimization of NACA wing with TACS `StaticProblem` class and an openMDAO `ExplicitComponent` optimizer written on top of it. An arbitrary number of thickness variables can be assigned using `capsGroup` attributes in the CSM file and the tacsAIM.
4. Shape optimization of NACA wing with rib and spar fraction shape variables. The ESP/CAPS design parameters move ribs and spars around inside the wing, with derivatives provided by ESP/CAPS. The coordinate derivatives df/dxi, df/dyi, df/dzi for each node in the mesh are written from TACS to a sensitivity file nastran_CAPS.sens in the tacsAIM work directory often called `capsStruct`. Then, the `tacs_aim.post_analysis()` call runs the postAnalysis on the root processor, reading in the derivatives to ESP/CAPS which knows the mesh derivatives dxi/dshapeDV, dyi/dshapeDV, dzi/dshapeDV and computes the total shape derivatives. 

2 offline unit tests were written to be able to verify the TACS->ESP/CAPS sensitivity writer and shape variable is correct. The other test verifies the thickness derivatives setup using `capsGroup` attributes. These tests are not run online at the moment since ESP/CAPS doesn't support complex numbers, and is not setup to install during the github workflow yet.

timryanb commented:
Thanks for the addition. I should have some time to review this in a couple of days.

We should make these tests runnable on the CI, even if complex isn't supported. This is the only way to guarantee that they will be checked as the code changes. We can add a skip condition for the tests if TACS is compiled in complex mode or if the ESP module is not available, see [here](https://github.com/smdogroup/tacs/blob/master/tests/element_tests/test_element_3d.py#L202) for example.

timryanb commented on tacs/pymeshloader.py:
```diff
@@ -894,3 +894,59 @@ def idMap(self, fromIDs, tacsIDDict):
                 return -1
+
+    @property
+    def struct_ids(self):
```
Are struct ID's the same thing as the node ID's? Can we call them Node ID's to be consistent with the rest of the code?

timryanb commented on tacs/pymeshloader.py:
```diff
@@ -894,3 +894,59 @@ def idMap(self, fromIDs, tacsIDDict):
+            struct_ids = self._get_all_struct_ids()
+        return struct_ids
+
+    def _get_local_struct_ids(self):
```
Use camel case for method names to stay consistent with the rest of code here (i.e. `_getLocalNodeIDs`)

sean-engelstad commented on tacs/pytacs.py:
```diff
@@ -794,6 +794,8 @@ def elemCallBack(dvNum, compID, compDescript, elemDescripts,
         self.x0 = self.assembler.createDesignVec()
         self.assembler.getDesignVars(self.x0)
 
+        return self
+
```
It allows you to do method cascading like this in my caps wing examples I added
`SPs = tacs_aim.fea_solver.initialize().createTACSProbsFromBDF()`

timryanb replied:
Could you make this a new method of `tacs_aim`? Basically, `SPs = tacs_aim.createTACSProbs()` and then have the`createTACSProbs` method call `self.fea_solver.initialize()` and then `self.fea_solver. createTACSProbsFromBDF()` for the user?

sean-engelstad replied:
Ok I'll do that and get rid of the self in initialize of pytacs

timryanb commented on tacs/pymeshloader.py:
```diff
@@ -894,3 +894,59 @@ def idMap(self, fromIDs, tacsIDDict):
+        bdf_nodes = [_ for _ in range(num_nodes)]
+        return self.getLocalNodeIDsFromGlobal(bdf_nodes, nastranOrdering=False)
+
+    def _struct_id_map(self):
```
`_NastranToLocalNodeIDMap`

timryanb commented on tacs/pymeshloader.py:
```diff
@@ -894,3 +894,59 @@ def idMap(self, fromIDs, tacsIDDict):
+        self._local_map = id_map
+        return id_map
+
+    def _get_all_struct_ids(self):
```
`_getAllLocalNodeIDs`

timryanb commented on tacs/problems/base.py:
```diff
@@ -1070,3 +1070,69 @@ def _addPressureFromPLOAD4(self, auxElems, loadInfo, scale=1.0):
                     auxElems, elemID, trac, faceIndex, nastranOrdering=True
                 )
+
+    def write_sensitivity_file(self, evalFuncs, sens_file, tacs_aim=None):
```
Use camelCase for method name

timryanb replied:
Use camelCase for `sensFile` & `tacsAIM` arguement

timryanb commented on tacs/problems/base.py:
```diff
@@ -1070,3 +1070,69 @@ def _addPressureFromPLOAD4(self, auxElems, loadInfo, scale=1.0):
+
+    def write_sensitivity_file(self, evalFuncs, sens_file, tacs_aim=None):
+        """
+        write an ESP/CAPS .sens file from the tacs aim
```
Use consistent docstring format (define input/outputs) see [here](https://github.com/smdogroup/tacs/blob/master/tacs/pytacs.py#L688-L732) for example

timryanb commented on tacs/problems/modal.py:
```diff
@@ -525,3 +525,6 @@ def writeSolution(self, outputDir=None, baseName=None, number=None, indices=None
                 modeName = baseName + "_%3.3d" % index
                 fileName = os.path.join(outputDir, modeName) + ".f5"
                 self.outputViewer.writeToFile(fileName)
+
+    def write_sensitivity_file(self, evalFuncs, sens_file):
+        return super().write_sensitivity_file(evalFuncs, sens_file)
```
This seems unnecessary, the method is already defined in `baseProblem`. Please remove

sean-engelstad replied:
Ok I will take it out, VScode auto-added those for me.

timryanb commented on tacs/problems/static.py:
```diff
@@ -1557,3 +1557,6 @@ def writeSolution(self, outputDir=None, baseName=None, number=None):
         if self.getOption("writeSolution"):
             base = os.path.join(outputDir, baseName) + ".f5"
             self.outputViewer.writeToFile(base)
+
+    def write_sensitivity_file(self, evalFuncs, sens_file):
+        return super().write_sensitivity_file(evalFuncs, sens_file)
```
see above

timryanb commented on tacs/problems/transient.py:
```diff
@@ -1388,3 +1388,6 @@ def writeSolution(self, outputDir=None, baseName=None, number=None, timeSteps=No
                 stepName = baseName + "_%3.3d" % timeStep
                 fileName = os.path.join(outputDir, stepName) + ".f5"
                 self.outputViewer.writeToFile(fileName)
+
+    def write_sensitivity_file(self, evalFuncs, sens_file):
+        return super().write_sensitivity_file(evalFuncs, sens_file)
```
see above

sean-engelstad commented:
The examples and unit tests are solid now although I do have a few comments on the unit tests for ESP/CAPS. However, the unit tests are still not active in the workflow yet and I am still working on that.
- The thickness derivatives don't really involve ESP/CAPS and therefore pass the complex step test down to machine precision.
- The shape derivatives involve mesh regeneration in ESP/CAPS which can change mesh topology and can only use FD not complex step test.
- The shape derivative test only passes with a reasonably refined mesh. I don't refine the mesh to full scale for the unit test so it only passes for dmass/dshape for one shape var and I can't add any additional analysis functions like ksfailure which are nonlinear.

sean-engelstad commented:
The results of the sizing optimization in `examples/caps_wing/3_sizing_optimization.py` ran on a single processor are the following. Starting from an initial equal thickness design the optimal panel thicknesses are selected to minimize mass w.r.t. aggregated ksfailure vm stress constraint. The new `TacsModel` class was used to setup the optimization problem as well as a new dedicated `TacsStaticComponent` OpenMDAO explicit component for optimization with OpenMDAO. The optimization history and final optimal thicknesses are shown below.

![opt_history](https://user-images.githubusercontent.com/89812682/218746304-e75f2e9f-6902-42dc-b195-4fa44a9511b5.png)

![Screenshot from 2023-02-14 07-57-22](https://user-images.githubusercontent.com/89812682/218746339-29e75df6-abe9-4368-b0f9-7cfc27009fe7.png)

timryanb commented:
> The examples and unit tests are solid now although I do have a few comments on the unit tests for ESP/CAPS. However, the unit tests are still not active in the workflow yet and I am still working on that.
> 
> * The thickness derivatives don't really involve ESP/CAPS and therefore pass the complex step test down to machine precision.
> * The shape derivatives involve mesh regeneration in ESP/CAPS which can change mesh topology and can only use FD not complex step test.
> * The shape derivative test only passes with a reasonably refined mesh. I don't refine the mesh to full scale for the unit test so it only passes for dmass/dshape for one shape var and I can't add any additional analysis functions like ksfailure which are nonlinear.

What weight are you using for the ksfailure in your test? If you use a really low value (~10), the function should become pretty linear

sean-engelstad commented:
I was using ksWeight=50.0 and safetyFactor=1.5 for the ksfailure function.

timryanb commented:
> I was using ksWeight=50.0 and safetyFactor=1.5 for the ksfailure function for the unit tests.

Use a lower value (10.0), you don't need to use anything too high since this is just a test. See my test [here](https://github.com/smdogroup/tacs/blob/master/tests/integration_tests/test_shell_plate_quad.py#L21) for example.

sean-engelstad commented:
Result of caps2tacs shape optimization with shape variables from ESP/CAPS design parameters rib_a1, rib_a2, spar_a1, spar_a2 which move the group of ribs and spars around. Notice the ribs were moved closer to the root of the wing which makes sense.

![Screenshot from 2023-02-19 18-37-37](https://user-images.githubusercontent.com/89812682/219982357-81faedd7-b398-4413-9488-cbb373d60e23.png)

![opt_history](https://user-images.githubusercontent.com/89812682/219982402-cb70e289-23ae-47ed-a419-3236685d2c30.png)

sean-engelstad commented:
The ESP/CAPS installation and unit tests have successfully been added to the github workflow now and ran in the tests that have just passed. The testflo line now has to run in tests directory not . because otherwise it will also try and run ESP/CAPS unit tests which I don't want it to do. The ESP/CAPS installation and related unit tests are only ran in real mode via ${{}} github workflow if statements. Therefore, the total runtime is not increased as real mode is still shorter than complex.

sean-engelstad commented:
> > I was using ksWeight=50.0 and safetyFactor=1.5 for the ksfailure function for the unit tests.
> 
> Use a lower value (10.0), you don't need to use anything too high since this is just a test. See my test [here](https://github.com/smdogroup/tacs/blob/master/tests/integration_tests/test_shell_plate_quad.py#L21) for example.

I included ksfailure with a ksWeight=10.0 for the caps_shape_derivatives test now and that passes

sean-engelstad commented:
> > The ESP/CAPS installation and unit tests have successfully been added to the github workflow now and ran in the tests that have just passed. The testflo line now has to run in tests directory not . because otherwise it will also try and run ESP/CAPS unit tests which I don't want it to do. The ESP/CAPS installation and related unit tests are only ran in real mode via ${{}} github workflow if statements. Therefore, the total runtime is not increased as real mode is still shorter than complex.
> 
> I'd like to have `testflo .` still work. Where are these tests? Can we add a `skipIF` to these tests to address this?

I added a workflow command to recursively find and delete any "test*" files in ESP/CAPS directory so no ESP/CAPS tests are included in the workflow. Therefore, I was able to change back to "testflo ." command. Also, I experienced an outlier failure of my caps_shape_derivatives test, which varies between 1e-4 and 1e-7 accuracy. I added a random seed to ensure this test doesn't fail the workflow due to random variability.

timryanb commented:
> > > The ESP/CAPS installation and unit tests have successfully been added to the github workflow now and ran in the tests that have just passed. The testflo line now has to run in tests directory not . because otherwise it will also try and run ESP/CAPS unit tests which I don't want it to do. The ESP/CAPS installation and related unit tests are only ran in real mode via ${{}} github workflow if statements. Therefore, the total runtime is not increased as real mode is still shorter than complex.
> > 
> > 
> > I'd like to have `testflo .` still work. Where are these tests? Can we add a `skipIF` to these tests to address this?
> 
> I added a workflow command to recursively find and delete any "test*" files in ESP/CAPS directory so no ESP/CAPS tests are included in the workflow. Therefore, I was able to change back to "testflo ." command. Also, I experienced an outlier failure of my caps_shape_derivatives test, which varies between 1e-4 and 1e-7 accuracy. I added a random seed to ensure this test doesn't fail the workflow due to random variability.

Thanks, sound good to me

sean-engelstad commented on tacs/problems/base.py:
```diff
@@ -1069,3 +1069,79 @@ def _addPressureFromPLOAD4(self, auxElems, loadInfo, scale=1.0):
+        num_nodes = self.meshLoader.bdfInfo.nnodes
+        node_ids = self.meshLoader.nodeIDs
+
+        if self.comm is None or self.comm.rank == 0:
```
Ok I will remove the None check since the `comm` can't be None in the `Base` class.

sean-engelstad commented on tacs/pymeshloader.py:
```diff
@@ -894,3 +894,58 @@ def idMap(self, fromIDs, tacsIDDict):
+        nastran_node - 1 => tacs_idx owned by this proc
+        """
+
+        if self.comm is None:
```
I removed the None check here and the method that no longer gets called

sean-engelstad commented on tacs/pymeshloader.py:
```diff
@@ -894,3 +894,58 @@ def idMap(self, fromIDs, tacsIDDict):
                 return -1
+
+    @property
+    def nodeIDs(self):
```
I changed this name to allLocalNodeIDs since after the comm=None case is removed it just would have called allLocalNodeIDs. See below conversation

timryanb commented on tacs/problems/base.py:
```diff
@@ -1069,3 +1069,79 @@ def _addPressureFromPLOAD4(self, auxElems, loadInfo, scale=1.0):
+        Parameters
+        ----------
+
+        evalFuncs : list of strings
```
That's right, eventually we should add a section to the tacs docs explaining how to use the caps2tacs interface. That can be done in a future PR though
